hir-transform/src/treeify.rs

```diff
@@ -327,15 +328,25 @@ impl RewritePass for Treeify {
                     }
                 }
 
-                // After treeification, the original subtree blocks cannot possibly be
-                // referenced by other blocks in the function, so remove all of them
-                detach_tree(b, function, &cfg);
+                // After treeification, the original subtree blocks cannot
+                // possibly be referenced by other blocks in the function, so
+                // add the subtree to the removal queue
+                //
+                // We don't want to detach the subtree (remove all the blocks)
+                // immediately because we might need to treeify a subtree of
+                // this subtree later in this loop
+                // see https://github.com/0xPolygonMiden/compiler/issues/194
+                detach_q.push_back(b);
 
                 // Mark the control flow graph as modified
                 changed = true;
             }
         }
 
+        for b in detach_q {
```

Comment:
Won't this cause us to hit the same bug, just at the end of treeification? I think it isn't evident here because you've also silenced the assertion that was previously raised, by ignoring when a detach of an already-detached block is attempted in `DataFlowGraph::detach_block`.

I think at a minimum it will be necessary to track which blocks we've already detached, and if we reach a subtree rooted at a block which we've already detached, then we skip `detach_tree` entirely. That has the same effect as your change to `DataFlowGraph::detach_block`, but is explicitly localized to just this specific operation where we know we might attempt to detach a block multiple times.

Reply:
> Won't this cause us to hit the same bug, just at the end of treeification? I think it isn't evident here because you've also silenced the assertion that was previously raised, by ignoring when a detach of an already-detached block is attempted in `DataFlowGraph::detach_block`.

Yes, you're right, it just silences the repeated detachment. I did this to resolve another issue with inserting a block after the detached block in the treefication (copying) process.